android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/BaseSearchTelemetry.kt

```diff
@@ -21,20 +21,53 @@ import mozilla.components.support.base.facts.collect
 import mozilla.components.support.base.log.logger.Logger
 import mozilla.components.support.ktx.kotlinx.coroutines.flow.filterChanged
 import org.json.JSONObject
+import java.io.File
+
+internal const val COLLECTION_NAME = "search-telemetry-v2"
 
 /**
  * Main configuration and functionality for tracking ads / web searches with specific providers.
  */
 abstract class BaseSearchTelemetry {
+    @VisibleForTesting
+    internal var providerList: List<SearchProviderModel>? = emptyList()
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    internal suspend fun initializeProviderList(rootStorageDirectory: File, readJson: () -> JSONObject) {
+        val serpRepo = SerpTelemetryRepository(rootStorageDirectory, COLLECTION_NAME)
+
+        val (cacheLastModified, cacheResponse) = serpRepo.fetchProvidersFromCache()
+        val localResponse = readJson()
+
+        if (cacheResponse.isEmpty() || cacheLastModified < localResponse.getString("timestamp").toULong()) {
+            providerList = serpRepo.parseLocalData(localResponse)
+        } else if (cacheLastModified > localResponse.getString("timestamp").toULong()) {
+            providerList = cacheResponse
+        }
+
+        val remoteResponse = serpRepo.fetchRemoteResponse()
+        if (remoteResponse.lastModified > cacheLastModified) {
+            serpRepo.writeToCache(remoteResponse)
```

Comment:
Should we also be setting the providerList here? I am not sure that we want to rely on a user restarting the app to be getting the latest data

Reply:
Makes sense! Thank you